rust/src/delta_datafusion.rs

```diff
@@ -263,29 +345,85 @@ impl TableProvider for delta::DeltaTable {
             self.object_store(),
         );
 
-        // TODO prune files based on file statistics and filter expressions
+        // The pruning predicate indicates which containers / files can be pruned.
+        // i.e. 'true' means the file is to be excluded
+        let files_to_prune = if !filters.is_empty() {
+            if let Some(predicate) = combine_filters(filters) {
+                let pruning_predicate = PruningPredicate::try_new(predicate, schema.clone())?;
+                pruning_predicate.prune(self)?
+            } else {
+                vec![false; self.get_state().files().len()]
+            }
+        } else {
+            vec![false; self.get_state().files().len()]
+        };
+
         let partitions = self
             .get_state()
             .files()
             .iter()
-            .map(|action| {
-                Ok(vec![PartitionedFile::new(
-                    action.path.clone(),
-                    action.size as u64,
-                )])
+            .zip(files_to_prune.into_iter())
+            .filter_map(|(action, prune_file)| {
+                if prune_file {
+                    return None;
+                }
```

Comment:
might be worth specializing this so we are not paying the penalty of iterating and checking prune file array when there is no file to prune. For example, the code below can be abstracted into a function, then we have two types of iterator loops that calls it. One iterator loop zips with files_to_prune, the other one simply just iterates through `get_state().files().`

Reply:
Makes sense! Updated it and to the opportunity to batch files by partition values. Thats not going to be optimal for many cases, but by default hopefully better than each file in its own (datafusion-)partition.